File: app/src/main/java/io/github/skywalkerdarren/simpleaccounting/adapter/ExchangeRateAdapter.java
package io.github.skywalkerdarren.simpleaccounting.adapter;

import android.content.Context;

import androidx.recyclerview.widget.RecyclerView;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.chad.library.adapter.base.listener.OnItemDragListener;

import java.util.Objects;

import io.github.skywalkerdarren.simpleaccounting.R;
import io.github.skywalkerdarren.simpleaccounting.base.BaseDraggableDataBindingAdapter;
import io.github.skywalkerdarren.simpleaccounting.databinding.ItemExchangeRateBinding;
import io.github.skywalkerdarren.simpleaccounting.model.AppRepository;
import io.github.skywalkerdarren.simpleaccounting.model.datasource.CurrencyDataSource;
import io.github.skywalkerdarren.simpleaccounting.model.entity.Currency;
import io.github.skywalkerdarren.simpleaccounting.model.entity.CurrencyInfo;
import io.github.skywalkerdarren.simpleaccounting.util.AppExecutors;

public class ExchangeRateAdapter extends BaseDraggableDataBindingAdapter<Currency, ItemExchangeRateBinding> {
    private AppRepository mRepository;

    public ExchangeRateAdapter(Context context) {
        super(R.layout.item_exchange_rate, null);
        mRepository = AppRepository.getInstance(new AppExecutors(), context);
        openLoadAnimation(BaseQuickAdapter.SLIDEIN_RIGHT);
        setOnItemDragListener(new OnItemDragListener() {
            @Override
            public void onItemDragStart(RecyclerView.ViewHolder viewHolder, int pos) {
            }

            @Override
            public void onItemDragMoving(RecyclerView.ViewHolder source, int from, RecyclerView.ViewHolder target, int to) {
                mRepository.changeCurrencyPosition(Objects.requireNonNull(getItem(from)), Objects.requireNonNull(getItem(to)));

            }

            @Override
            public void onItemDragEnd(RecyclerView.ViewHolder viewHolder, int pos) {
            }
        });
    }

    @Override
    protected void convert(ItemExchangeRateBinding binding, Currency item) {
        binding.setData(item);
        mRepository.getCurrencyInfo(item.getName(), new CurrencyDataSource.LoadCurrencyInfoCallback() {
            @Override
            public void onCurrencyInfoLoaded(CurrencyInfo info) {
                if (info.getFullNameCN() != null) {
                    info.setFullName(info.getFullNameCN());
                }
                if (info.getFullName() == null) {
                    info.setFullName(info.getName());
                }
                binding.setInfo(info);
            }

            @Override
            public void onDataUnavailable() {

            }
        });
    }
}

File: app/src/main/java/io/github/skywalkerdarren/simpleaccounting/adapter/ExchangeRateAdapter.kt
package io.github.skywalkerdarren.simpleaccounting.adapter

import android.animation.Animator
import android.view.animation.AccelerateDecelerateInterpolator
import androidx.recyclerview.widget.RecyclerView
import com.chad.library.adapter.base.listener.OnItemDragListener
import com.chad.library.adapter.base.viewholder.BaseDataBindingHolder
import io.github.skywalkerdarren.simpleaccounting.R
import io.github.skywalkerdarren.simpleaccounting.adapter.diff.DefaultDiff
import io.github.skywalkerdarren.simpleaccounting.base.BaseDraggableDataBindingAdapter
import io.github.skywalkerdarren.simpleaccounting.databinding.ItemExchangeRateBinding
import io.github.skywalkerdarren.simpleaccounting.model.entity.CurrencyAndInfo
import io.github.skywalkerdarren.simpleaccounting.view_model.DiscoveryViewModel

class ExchangeRateAdapter(private val viewModel: DiscoveryViewModel) :
        BaseDraggableDataBindingAdapter<CurrencyAndInfo, ItemExchangeRateBinding>(R.layout.item_exchange_rate, null) {
    var isDrag = false
        private set

    override fun convert(holder: BaseDataBindingHolder<ItemExchangeRateBinding>, item: CurrencyAndInfo) {
        holder.dataBinding?.data = item.currency
        holder.dataBinding?.info = item.currencyInfo
    }

    override fun startAnim(anim: Animator, index: Int) {
        anim.duration = 300
        anim.interpolator = AccelerateDecelerateInterpolator()
        anim.start()
    }

    fun setNewList(currencies: List<CurrencyAndInfo>) {
        setDiffNewData(currencies.toMutableList())
    }

    init {
        setDiffCallback(DefaultDiff())
        draggableModule.setOnItemDragListener(object : OnItemDragListener {
            override fun onItemDragStart(viewHolder: RecyclerView.ViewHolder, pos: Int) {
                isDrag = true
            }

            override fun onItemDragMoving(source: RecyclerView.ViewHolder, from: Int, target: RecyclerView.ViewHolder, to: Int) {
                viewModel.changeCurrency(getItem(from).currency, getItem(to).currency)
            }

            override fun onItemDragEnd(viewHolder: RecyclerView.ViewHolder, pos: Int) {
                isDrag = false
            }
        })
    }
}